### ISSUE TITLE
Always restart stream on error and avoid ANR

### ISSUE THREAD

**Issue by jinfanzon** [google/ExoPlayer#7297]
### Searched documentation and issues
The main guide I followed to try to always restart HLS stream on error is the blog written by @AquilesCanta:
https://medium.com/google-exoplayer/load-error-handling-in-exoplayer-488ab6908137

### Question
I'm developing an app that it always has to restart the HLS stream when an error occurs. The app runs 24/7 streaming a live channel. I've tried to use the default error handling policy, also tried the proposed blog policy "Do not surface load errors due to lack of network connectivity" :

> public final class CustomPolicy extends DefaultLoadErrorHandlingPolicy { 

     @Override
     public long getRetryDelayMsFor(
         int dataType,
         long loadDurationMs,
         IOException exception,
         int errorCount) {
             // Replace NoConnectivityException with the corresponding
            // exception for the used DataSource.
           if (exception instanceof NoConnectivityException) {
               return 5000; // Retry every 5 seconds.
            } else {
               return C.TIME_UNSET; // Anything else is surfaced.
            }
      }
     @Override
      public int getMinimumLoadableRetryCount(int dataType) {
          return Integer.MAX_VALUE;
      }}
Changing NoConnectivityException issue for HttpDataSource.HttpDataSourceException.
Without any successful result, If I leave the app running (on a Chinese android set-top-box) all day at some point it gets tilt. 
Together with the error policy, I've also have implemented Player.EventListener overriding OnPlayer error and there I release the player and create it (re-prepare) again with the mediasource.

I've observed in the tombstones that there is an ANR on the mediacodec:

> _pid: 179, tid: 12767, name: Binder:179_6  >>> media.codec <<<
signal 11 (SIGSEGV), code 1 (SEGV_MAPERR), fault addr 0xdc
    r0 00000000  r1 ae480000  r2 00000000  r3 000759a8
    r4 b0d887c0  r5 ae480000  r6 00000000  r7 acd6381c
    r8 00000000  r9 0000004a  sl 00000000  fp 0000004a
    ip 00000000  sp acd634a0  lr 0000004a  pc b0a1c864  cpsr 60000010
    d0  0000000000000000  d1  0000000000000000
    d2  0000000000000000  d3  0000000000000000
    d4  0000000000000000  d5  0000000000000000
    d6  0000000000000000  d7  0000000000000000
    d8  000000001ad310c0  d9  0000000000000000
    d10 0000000000000000  d11 0000000000000000
    d12 0000000000000000  d13 0000000000000000
    d14 0000000000000000  d15 0000000000000000
    d16 0000000000000001  d17 0000000000000307
    d18 b13e1d63b13e1d5b  d19 b13e1d6fb13e1d6b
    d20 0000000000000000  d21 0000000000000000
    d22 0000000000000000  d23 0000000000000000
    d24 0000000000000000  d25 0000000000000000
    d26 0000000000000000  d27 0000000000000000
    d28 0000000000000000  d29 0000000000000000
    d30 0000000000000000  d31 0000000000000000
    scr 80000010
backtrace:
    #00 pc 00016864  /system/lib/librk_vpuapi.so (_ZN7android13Rk_AvcDecoder11getOneFrameEPhPjRx+240)
    #01 pc 000176a8  /system/lib/librk_vpuapi.so (_ZN7android13Rk_AvcDecoder24pv_rkavcdecoder_oneframeEPhPjS1_S2_Rx+652)
    #02 pc 000183d0  /system/lib/librk_vpuapi.so (dec_oneframe_class_RkAvcDecoder+36)
    #03 pc 000131a5  /system/lib/librk_vpuapi.so (_ZN6VpuApi8RkDecodeEP15VpuCodecContextP10RkDecInputP11RkDecOutput+140)
    #04 pc 0001367d  /system/lib/librk_vpuapi.so (_ZN6VpuApi12codec_threadEPv+284)
    #05 pc 00046d33  /system/lib/libc.so (_ZL15__pthread_startPv+22)
    #06 pc 00019aad  /system/lib/libc.so (__start_thread+6)_`

So to summarize, my question is which is the correct error handle policy I should implement to always restart on error? What code implementation should I have on "OnPLayerError"? Will the proposed solution work despite an ANR? should this ANR be a bug?

### A full bug report captured from the device
I'm sending the bugreport, please refer to tombstone_00 file.

### Link to test content
N/A

**Comment by christosts** [comment]
I see in the bugreport logs that ExoPlayer.release() is called, so I assume you code is working as intended, that is to stop, release and re-create the player.

I think the ANR is unrelated to the LoadErrorHandlingPolicy and is probably device-specific. Have you tried this on another device? Does the ANR happen there too? On the device that happens, what is the frequency? Can you simulate "errors" in order to trigger the stop/release/re-create flow and see how frequently the ANR occurs?

We've seen ANRs from MediaCode.release(), but the root cause can be the underlying device implementation. There's not much the ExoPlayer can do there. 

I can offer some assistance to debug this. If you use the dev-v2 branch, there's currently an experimental method to set a timeout when ExoPlayer calls MediaCodec.release() - if it times out, the player will raise an error. You can set the timeout when building an ExoPlayer instance with the `ExoPlayer.Builder` by calling [experimental_setReleaseTimeoutMs](https://github.com/google/ExoPlayer/blob/f921d0d3e59264c719a9a74a6efc07d58fdcf9e0/library/core/src/main/java/com/google/android/exoplayer2/ExoPlayer.java#L226). Note, this does not fix nor address the issue, but only helps verify if the root cause is MediaCodec.release() being blocked.

Marking this as device-specific, unless there are data that this occurs on other devices too.

**Comment by jinfanzon** [comment]
Hi @christosts  - This happened in multiple chinese android tv box, They are different brands, but beneath they might run the same hardware/firmware. I'll try to set experimental_setReleaseTimeoutMs, though I'm not sure I understand the expected behaviour/debug output.

Kind regards!

**Comment by christosts** [comment]
`They are different brands, but beneath they might run the same hardware/firmware`

From the logs of the shared bug-report, looks like the device is based on Rockchip hardware. Look for variant of the `rockschip` string printed in the logs of the other devices to identify if they are built with similar hardware.

`I'll try to set experimental_setReleaseTimeoutMs, though I'm not sure I understand the expected behaviour/debug output.`

ExoPlayer.release() will raise a TimeoutException if MediaCodec.release() did not complete within the time you configured. When calling ExoPlayer.release() from your app, catch any exception (and log it for debugging). If the exception is indeed raised, this will verify it is an issue coming from MediaCodec.release(). Setting the timeout would only prevent the ANR since you can catch the exception and continue operating the app. But it will not effectively solve the problem. If the MediaCodec is in a bad state, it is very unlikely that you will be able to re-start playback with a new player instance.
